karrot/places/tests/test_place_types_api.py

```diff
@@ -0,0 +1,158 @@
+        response = self.client.post('/api/place-types/', self.place_type_data(), format='json')
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_cannot_create_for_group_if_not_member(self):
+        self.client.force_login(user=self.member)
+        other_group = GroupFactory()
+        response = self.client.post(
+            '/api/place-types/', self.place_type_data({
+                'group': other_group.id,
+            }), format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_can_create_place_for_custom_type(self):
+        self.client.force_login(user=self.member)
+        response = self.client.post('/api/place-types/', self.place_type_data(), format='json')
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED, response.data)
+        place_response = self.client.post(
+            '/api/places/', self.place_data({
+                'place_type': response.data['id'],
+            }), format='json'
+        )
+        self.assertEqual(place_response.status_code, status.HTTP_201_CREATED, place_response.data)
+
+    def test_can_modify(self):
+        self.client.force_login(user=self.member)
+        response = self.client.patch(
+            f'/api/place-types/{self.place_type.id}/', {
+                'icon': 'fas fa-changed',
+            }, format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
+
+    def test_cannot_modify_if_not_editor(self):
+        self.client.force_login(user=self.non_editor_member)
+        response = self.client.patch(
+            f'/api/place-types/{self.place_type.id}/', {
+                'icon': 'fas fa-changed',
+            }, format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_cannot_modify_group(self):
+        self.client.force_login(user=self.member)
+        other_group = GroupFactory()
+        response = self.client.patch(
+            f'/api/place-types/{self.place_type.id}/', {
+                'group': other_group.id,
+            }, format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+
+    def test_can_delete(self):
+        self.client.force_login(user=self.member)
+        response = self.client.delete(f'/api/place-types/{self.place_type.id}/')
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT, response.data)
+
+    def test_cannot_delete_with_places(self):
+        self.client.force_login(user=self.member)
+        place = PlaceFactory(place_type=self.place_type)
+        response = self.client.delete(f'/api/place-types/{self.place_type.id}/')
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+        # make sure we can delete it if we get rid of the activity
```

Comment:
typo: activity -> place

Reply:
Although places cannot be deleted, just archived! So I wonder if it actually makes sense to delete place types, except newly created ones.

Reply:
Yup, maybe doesn't actually get used in the API, although I would just keep it possible.